src/components/structures/TimelinePanel.js

```diff
@@ -1119,6 +1130,73 @@ const TimelinePanel = createReactClass({
         };
     },
 
+    /**
+     * Check for undecryptable messages that were sent while the user was not in
+     * the room.
+     *
+     * @param {Array<MatrixEvent>} events The timeline events to check
+     * @param {Room} room The room that the events are in
+     *
+     * @return {Number} The index within `events` of the event after the most recent
+     * undecryptable event that was sent while the user was not in the room.  If no
+     * such events were found, then it returns 0.
+     */
+    _checkForPreJoinUISI: function(events) {
+        const room = this.props.timelineSet.room;
+
+        if (events.length === 0 || !room ||
+            !MatrixClientPeg.get().isRoomEncrypted(room.roomId)) {
+            return 0;
+        }
+
+        const userId = MatrixClientPeg.get().credentials.userId;
+
+        let i;
+        let userMembership = "leave";
+        // find the last event that is in a timeline (events may not be in a
+        // timeline if they have not been sent yet) and get the user's membership
+        // at that point.
+        for (i = events.length - 1; i >= 0; i--) {
+            const timeline = room.getTimelineForEvent(events[i].getId());
+            if (timeline) {
+                const userMembershipEvent =
+                      timeline.getState(EventTimeline.FORWARDS).getMember(userId);
+                userMembership = userMembershipEvent ? userMembershipEvent.membership : "leave";
+                const timelineEvents = timeline.getEvents();
+                for (let j = timelineEvents.length - 1; j >= 0; j--) {
+                    const event = timelineEvents[j];
+                    if (event.getId() === events[i].getId()) {
+                        break;
+                    } else if (event.getStateKey() === userId
+                        && event.getType() === "m.room.member") {
+                        const prevContent = event.getPrevContent();
+                        userMembership = prevContent.membership || "leave";
+                    }
+                }
+                break;
+            }
+        }
+
+        // now go through the rest of the events and find the first undecryptable
+        // one that was sent when the user wasn't in the room
+        for (; i >= 0; i--) {
+            const event = events[i];
+            if (event.getStateKey() === userId
+                && event.getType() === "m.room.member") {
+                const prevContent = event.getPrevContent();
+                userMembership = prevContent.membership || "leave";
+            } else if (userMembership === "leave" &&
+                       (event.isDecryptionFailure() || event.isBeingDecrypted())) {
```

Comment:
One result of moving this to `TimelinePanel` is that I had to make it also hide things that are in the process of being decrypted (which is probably a wrong thing to do), rather than just things that failed to decrypt, presumably because it hasn't had enough time to finish trying to decrypt.  It would be nice if there was a better way to do this, but I can't see anything better.

Reply:
Hmm, not sure I follow why that's only an issue up here in `TimelinePanel`, I would have assumed it would apply in either component...

What is the user experience with and without including events being decrypted here?

Reply:
I'm not entirely sure why it only happens when it's in `TimelinePanel` and not in `MessagePanel`.  My best guess is that it's a timing issue -- by the time `MessagePanel` gets called, the message decryption has completed, whereas in `TimelinePanel`, it's still trying to decrypt.

If I don't include `isBeingDecrypted`, then it will just show all the events, without dropping the pre-join UISIs.